### PR TITLE
[WIP] Introduce BackendSelectId dispatch key

### PR DESCRIPTION
**Pull request by izdeby**

### CODE REVIEW

**Comment by kostmo** [comment]
## :pill: CircleCI build failures summary and remediations
As of commit 7164e2a9:


* **2/2** failures introduced in this PR


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=7164e2a98a1a70cb475c6147fb344ab43642e0c4).

### :detective: 2 new failures recognized by patterns
The following build failures do not appear to be due to upstream breakage:
#### [See CircleCI build](https://circleci.com/gh/pytorch/pytorch/4329093) pytorch_linux_xenial_py3_6_gcc5_4_ge_config_simple_test (1/2)
**Step:** "Test" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=90094684) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=90094684))
<details>
<summary>
<code>Jan 24 00:21:03 RuntimeError: test_jit_simple failed!</code>
</summary>

```
Jan 24 00:21:03 Ran 2352 tests in 75.983s 
Jan 24 00:21:03  
Jan 24 00:21:03 FAILED (errors=2, skipped=79, expected failures=1) 
Jan 24 00:21:03  
Jan 24 00:21:03 Generating XML reports... 
Jan 24 00:21:03 Traceback (most recent call last): 
Jan 24 00:21:03   File "test/run_test.py", line 456, in <module> 
Jan 24 00:21:03     main() 
Jan 24 00:21:03   File "test/run_test.py", line 449, in main 
Jan 24 00:21:03     raise RuntimeError(message) 
Jan 24 00:21:03 RuntimeError: test_jit_simple failed! 
Jan 24 00:21:03 + cleanup 
Jan 24 00:21:03 + retcode=1 
Jan 24 00:21:03 + set +x 
Jan 24 00:21:03 =================== sccache compilation log =================== 
Jan 24 00:21:03 =========== If your build fails, please take a look at the log above for possible reasons =========== 
Jan 24 00:21:03 Compile requests                 2 
Jan 24 00:21:03 Compile requests executed        0 
Jan 24 00:21:03 Cache hits                       0 
Jan 24 00:21:03 Cache misses                     0 
Jan 24 00:21:03 Cache timeouts                   0 
```

</details>

#### [See CircleCI build](https://circleci.com/gh/pytorch/pytorch/4329091) pytorch_linux_xenial_py3_6_gcc5_4_ge_config_legacy_test (2/2)
**Step:** "Test" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=90095190) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=90095190))
<details>
<summary>
<code>Jan 24 00:21:39 RuntimeError: test_jit_legacy failed!</code>
</summary>

```
Jan 24 00:21:39 Ran 2352 tests in 136.808s 
Jan 24 00:21:39  
Jan 24 00:21:39 FAILED (errors=2, skipped=62, expected failures=1) 
Jan 24 00:21:39  
Jan 24 00:21:39 Generating XML reports... 
Jan 24 00:21:39 Traceback (most recent call last): 
Jan 24 00:21:39   File "test/run_test.py", line 456, in <module> 
Jan 24 00:21:39     main() 
Jan 24 00:21:39   File "test/run_test.py", line 449, in main 
Jan 24 00:21:39     raise RuntimeError(message) 
Jan 24 00:21:39 RuntimeError: test_jit_legacy failed! 
Jan 24 00:21:39 =================== sccache compilation log =================== 
Jan 24 00:21:39 + cleanup 
Jan 24 00:21:39 + retcode=1 
Jan 24 00:21:39 + set +x 
Jan 24 00:21:40 =========== If your build fails, please take a look at the log above for possible reasons =========== 
Jan 24 00:21:40 Compile requests               260 
Jan 24 00:21:40 Compile requests executed        0 
Jan 24 00:21:40 Cache hits                       0 
Jan 24 00:21:40 Cache misses                     0 
Jan 24 00:21:40 Cache timeouts                   0 
```

</details>



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

**Comment by ezyang** [comment]
One nit: I think we should probably just call it `DispatchKey::BackendSelect`; I know all of the other keys have `TensorId` in their name, but that naming is kind of out of date now that we renamed `TensorTypeId` to `DispatchKey`

**Review by gchanan — CHANGES_REQUESTED** [review]
please add a PR description.

For example:
what does BackendSelectId do?
why do you need it?

File: aten/src/ATen/native/BackendSelectRegister.cpp [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+#include <ATen/core/op_registration/op_registration.h>
+#include <c10/core/TensorOptions.h>
+
+// TODO:
+// 1) add all factory functions
+// 2) autogen this
+
+namespace at {
+namespace native {
+  Tensor empty_backend(IntArrayRef size, const TensorOptions & options, c10::optional<MemoryFormat> memory_format) {
```
Comment (gchanan) [inline aten/src/ATen/native/BackendSelectRegister.cpp]: is there a reason these aren't in an anonymous namespace (with the "plain" name, i.e. without the backend suffix)?

File: aten/src/ATen/native/BackendSelectRegister.cpp [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+    DispatchKey key = options.computeDispatchKey();
+    static auto op = c10::Dispatcher::singleton().findSchemaOrThrow("aten::empty", "memory_format");
+    return op.callUnboxedWithDispatchKey<Tensor, IntArrayRef, const TensorOptions &, c10::optional<MemoryFormat>>(key, size, options, memory_format);
+  }
+
+  Tensor empty_strided_backend(IntArrayRef size, IntArrayRef stride, const TensorOptions & options) {
```
Comment (gchanan) [inline aten/src/ATen/native/BackendSelectRegister.cpp]: this indentation (inside the namespace) doesn't follow other examples I checked in this folder.

File: aten/src/ATen/native/BackendSelectRegister.cpp [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+    return op.callUnboxedWithDispatchKey<Tensor, IntArrayRef, IntArrayRef, const TensorOptions &>(key, size, stride, options);
+  }
+
+  Tensor empty_affine_quantized_backend(IntArrayRef size, const TensorOptions & options, double scale, int64_t zero_point, c10::optional<MemoryFormat> memory_format) {
+    DispatchKey key = options.computeDispatchKey();
+    static auto op = c10::Dispatcher::singleton().findSchemaOrThrow("aten::_empty_affine_quantized", "");
+    return op.callUnboxedWithDispatchKey<Tensor, IntArrayRef, const TensorOptions &, double, int64_t, c10::optional<MemoryFormat>>(key, size, options, scale, zero_point, memory_format);
+  }
+
+  Tensor empty_per_channel_affine_quantized_quantized_backend(IntArrayRef size, const Tensor & scales, const Tensor & zero_points, int64_t axis, const TensorOptions & options, c10::optional<MemoryFormat> memory_format) {
```
Comment (gchanan) [inline aten/src/ATen/native/BackendSelectRegister.cpp]: why this set of operators?  I'm guessing this is supposed to be some "base" factory functions that others will build on, but I don't see how this one fits that example -- look at the CPU implementation of this function to see why.

**Review by gchanan — CHANGES_REQUESTED** [review]

File: aten/src/ATen/native/BackendSelectRegister.cpp [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+    return op.callUnboxedWithDispatchKey<Tensor, IntArrayRef, const TensorOptions &, c10::optional<MemoryFormat>>(key, size, options, memory_format);
+  }
+
+  Tensor empty_strided_backend(IntArrayRef size, IntArrayRef stride, const TensorOptions & options) {
+    DispatchKey key = options.computeDispatchKey();
+    static auto op = c10::Dispatcher::singleton().findSchemaOrThrow("aten::empty_strided", "");
+    return op.callUnboxedWithDispatchKey<Tensor, IntArrayRef, IntArrayRef, const TensorOptions &>(key, size, stride, options);
+  }
+
+  Tensor empty_affine_quantized_backend(IntArrayRef size, const TensorOptions & options, double scale, int64_t zero_point, c10::optional<MemoryFormat> memory_format) {
```
Comment (gchanan) [inline aten/src/ATen/native/BackendSelectRegister.cpp]: I'm a little surprised to see TensorOptions here; I'm guessing this is an intermediate state, but a comment to that effect would be helpful.

File: c10/core/TensorOptions.h [hunk 5]
```diff
@@ -337,7 +337,7 @@ struct C10_API TensorOptions {
 
   // Resolves the tensor type set specified by the current construction axes.
   DispatchKeySet key_set() const noexcept {
-    return DispatchKeySet(computeDispatchKey()).add(DispatchKey::VariableTensorId);
+    return DispatchKeySet(computeDispatchKey()).add(DispatchKey::VariableTensorId).add(DispatchKey::BackendSelectId);
```
Comment (gchanan) [inline c10/core/TensorOptions.h]: for my own education: is this the only place where we add default TensorTypeIds?  Seems a bit strange to only be on TensorOptions.

Comment (izdeby) [reply c10/core/TensorOptions.h]: Yes, there are: 

https://github.com/pytorch/pytorch/blob/master/c10/core/impl/LocalDispatchKeySet.cpp#L29-L33

https://github.com/pytorch/pytorch/blob/master/c10/core/TensorImpl.cpp#L61

File: aten/src/ATen/native/BackendSelectRegister.cpp (hunk 4 above)
Comment (izdeby) [reply aten/src/ATen/native/BackendSelectRegister.cpp]: Once TO are gone, these guys will look something like this: 
`
Tensor full_like(const Tensor & self, Scalar fill_value, c10::optional<c10::ScalarType> dtype, c10::optional<c10::Layout> layout, c10::optional<c10::Device> device, c10::optional<bool> pin_memory, c10::optional<MemoryFormat> memory_format) {
  DispatchKey key = TensorOptions.calculate_dispatch_key(dtype, layout, device);
  static auto op = c10::Dispatcher::singleton().findSchemaOrThrow("aten::full_like", "dtype");
  return op.callUnboxedWithDispatchKey<Tensor, const Tensor &, Scalar, c10::optional<c10::ScalarType>, c10::optional<c10::Layout>, c10::optional<c10::Device>, c10::optional<bool>, c10::optional<MemoryFormat>>(key, self, fill_value, dtype, layout, device, pin_memory, memory_format);
}
`

**Pull request closed without merging**